oauth2/src/test/java/com/linecorp/armeria/client/auth/oauth2/OAuth2ClientCredentialsGrantTest.java

```diff
@@ -0,0 +1,209 @@
+                    .withAuthorizedClient("test_server", "server_secret")
+                    .withClientToken("test_client", TOKEN));
+        }
+    };
+
+    private final ServerRule resourceServerRule = new ServerRule(false) {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            final WebClient introspectClient = WebClient.of(authServerRule.httpUri());
+            sb.service("/resource-read-write/",
+                       OAuth2TokenIntrospectionAuthorizer.builder(introspectClient, "/introspect/token/")
+                                                         .realm("protected resource read-write")
+                                                         .accessTokenType("Bearer")
+                                                         .clientBasicAuthorization(() -> SERVER_CREDENTIALS)
+                                                         .permittedScope("read", "write")
+                                                         .build().asAuthService(SERVICE));
+            sb.service("/resource-read/",
+                       OAuth2TokenIntrospectionAuthorizer.builder(introspectClient, "/introspect/token/")
+                                                         .realm("protected resource read")
+                                                         .accessTokenType("Bearer")
+                                                         .clientBasicAuthorization(() -> SERVER_CREDENTIALS)
+                                                         .permittedScope("read")
+                                                         .build().asAuthService(SERVICE));
+            sb.service("/resource-read-write-update/",
+                       OAuth2TokenIntrospectionAuthorizer.builder(introspectClient, "/introspect/token/")
+                                                         .realm("protected resource read-write-update")
+                                                         .accessTokenType("Bearer")
+                                                         .clientBasicAuthorization(() -> SERVER_CREDENTIALS)
+                                                         .permittedScope("read", "write", "update")
+                                                         .build().asAuthService(SERVICE));
+        }
+    };
+
+    @Test
+    public void testOk() throws Exception {
+        try (Server resourceServer = resourceServerRule.start()) {
+
+            final WebClient authClient = WebClient.of(authServerRule.httpUri());
+            final OAuth2ClientCredentialsGrant grant = OAuth2ClientCredentialsGrant
+                    .builder(authClient, "/token/client/")
+                    .clientBasicAuthorization(() -> CLIENT_CREDENTIALS).build();
+
+            final WebClient client = WebClient.builder(resourceServerRule.httpUri())
+                                              .decorator(OAuth2Client.newDecorator(grant))
```

Comment:
`OAuth2Client` looks nice! However, it has to create `WebClient` every time for different resource owners.

How about adding a new method such as `signedRequest(HttpRequest)` to OAuth2ClientCredentialsGrant? Then users can set an HttpRequest that has `AUTHORIZATION` header to the existing WebClient without creating a new one. 

```java
class OAuth2ClientCredentialsGrant {
    CompletableFuture<HttpRequest> signedRequest(HttpRequest req) {
        return getAccessToken().thenApply(accessToken -> {
            return req.withHeaders(/* set AUTHORIZATION header */);
        });
    }
}

HttpRequest request = HttpRequest.of(HttpMethod.GET, "/resource-read-write/");
CompletableFuture<HttpRequest> signedRequest = grant.signRequest(request);
AggregatedHttpResponse response = 
        signedRequest.thenCompose(signed -> client.execute(signed).aggregate()).join();
```

Reply:
Good idea! Added this as `withAuthorization` to `OAuth2AuthorizationGrant` interface and supplied with the unit testing.